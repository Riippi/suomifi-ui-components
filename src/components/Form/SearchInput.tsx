import React, { Component } from 'react';
import { BaseTextInput, TextInputProps } from './TextInput';
import { logger } from '../../utils/logger';

export class SearchInput extends Component<TextInputProps> {
  render() {
    const { placeholder, ...passProps } = this.props;

    if (!!placeholder) {
      logger.warn(
        'Manually set placeholder attribute is not supported, use labelText instead',
      );
    }
    const labelAsPlaceholder = this.props.labelMode === 'hidden';
    const props = {
      placeholder: labelAsPlaceholder ? this.props.labelText : undefined,
    };

    return <BaseTextInput {...passProps} {...props} />;
  }
}
