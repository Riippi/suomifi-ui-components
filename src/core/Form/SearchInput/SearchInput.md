```js
import { SearchInput } from 'suomifi-ui-components';
<>
  <SearchInput
    onBlur={event => console.log(event.target.value)}
    labelText="Search..."
    labelMode="hidden"
  />

  <SearchInput
    onBlur={event => console.log(event.target.value)}
    labelText="Search..."
    labelMode="visible"
    placeholder="Any value (see console)"
  />
</>;
```
