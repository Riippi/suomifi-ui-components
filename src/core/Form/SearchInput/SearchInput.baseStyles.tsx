import { css } from 'styled-components';
import { suomifiTheme } from '../../theme';
import { SearchInputProps } from './SearchInput';
import { math } from 'polished';

export const baseStyles = ({ theme = suomifiTheme }: SearchInputProps) => css`
  & .fi-search-input {
    &_input-container {
      position: relative;
    }
    &_input {
      padding-right: ${math(`${theme.spacing.m} * 2 + ${theme.spacing.s}`)};
    }
    &_icon {
      position: absolute;
      top: 50%;
      right: ${theme.spacing.m};
      margin-top: -0.5em;
    }
  }
`;
