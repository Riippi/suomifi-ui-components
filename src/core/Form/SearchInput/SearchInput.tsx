import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { InternalTokensProp } from '../../theme';
import { withSuomifiDefaultProps } from '../../theme/utils';
import { baseStyles } from './SearchInput.baseStyles';
import { baseStyles as inputBaseStyles } from '../TextInput/TextInput.baseStyles';
import { SearchInput as CompSearchInput } from '../../../components/Form/SearchInput';
import { TextInputProps, textInputClassNames } from '../TextInput/TextInput';
import { Icon } from '../../Icon/Icon';
import classnames from 'classnames';
import { Omit } from '../../../utils/typescript';

const baseClassName = 'fi-search-input';
const inputContainerBaseClassName = `${baseClassName}_input-container`;
const inputBaseClassName = `${baseClassName}_input`;
const iconBaseClassName = `${baseClassName}_icon`;

export interface SearchInputProps extends Omit<TextInputProps, 'variant'> {}

const StyledTextInput = styled(
  ({
    tokens,
    className,
    labelTextProps = { className: undefined },
    inputContainerProps = { className: inputContainerBaseClassName },
    inputClassName,
    ...passProps
  }: TextInputProps & InternalTokensProp) => (
    <CompSearchInput
      {...passProps}
      labelTextProps={{
        ...labelTextProps,
        className: classnames(
          labelTextProps.className,
          textInputClassNames.labelParagraph,
        ),
      }}
      inputContainerProps={{
        ...inputContainerProps,
        className: classnames(
          inputContainerProps.className,
          textInputClassNames.inputContainer,
        ),
      }}
      {...passProps}
      className={classnames(className, baseClassName)}
      inputClassName={classnames(inputClassName, inputBaseClassName)}
    />
  ),
)`
  ${props => inputBaseStyles(props)}
  ${props => baseStyles(props)}
`;

/**
 * <i class="semantics" />
 * Use for user inputting search text
 */
export class SearchInput extends Component<SearchInputProps> {
  render() {
    return (
      <StyledTextInput {...withSuomifiDefaultProps(this.props)}>
        <Icon icon="search" className={iconBaseClassName} />
      </StyledTextInput>
    );
  }
}
