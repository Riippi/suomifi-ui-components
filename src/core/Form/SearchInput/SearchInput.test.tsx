import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../../utils/test/axe';

import { SearchInput } from './SearchInput';

const TestSearchInput = (
  <SearchInput labelText="Test search input" data-testid="textinput" />
);

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(TestSearchInput);
  const { container } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestSearchInput));
