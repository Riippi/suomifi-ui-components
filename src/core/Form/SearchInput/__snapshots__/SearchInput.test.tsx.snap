// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c4::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5 {
  display: inline-block;
  vertical-align: baseline;
}

.c1.fi-text-input {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  display: inline-block;
  width: 290px;
}

.c1 .fi-text-input_label-p {
  margin-bottom: 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,16%);
}

.c1 .fi-text-input_input-element-container > input:focus {
  outline-color: hsl(23,82%,53%);
  outline-width: 4px;
  outline-offset: 2px;
}

.c1 .fi-text-input_input-element-container:focus-within {
  outline: 0;
  position: relative;
}

.c1 .fi-text-input_input-element-container:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,55%);
  z-index: 9999;
}

.c1 .fi-text-input_input-element-container:focus-within:not(:focus-visible):after {
  content: none;
}

.c1 .fi-text-input_input-element-container:focus-within > input:focus {
  outline: none;
}

.c1 .fi-text-input_statusText_container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 .fi-text-input_statusText_container .fi-text-input_statusText {
  margin-top: 5px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
}

.c1 .fi-text-input_hintText {
  display: block;
  color: hsl(0,0%,16%);
  margin-bottom: 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 .fi-text-input_input {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,80%);
  border-radius: 2px;
  line-height: 1;
  background-color: hsl(0,0%,100%);
  min-width: 40px;
  width: 100%;
  min-height: 40px;
  padding-left: 10px;
}

.c1 .fi-text-input_input::-webkit-input-placeholder {
  font-style: italic;
}

.c1 .fi-text-input_input::-moz-placeholder {
  font-style: italic;
}

.c1 .fi-text-input_input:-ms-input-placeholder {
  font-style: italic;
}

.c1 .fi-text-input_input::placeholder {
  font-style: italic;
}

.c1 .fi-text-input_input:focus {
  box-shadow: 0 1px 2px 0 rgba(0,53,122,0.1) inset;
}

.c1.fi-text-input_with-icon .fi-text-input_input-element-container {
  position: relative;
}

.c1.fi-text-input_with-icon .fi-text-input_input {
  padding-right: 40px;
}

.c1.fi-text-input_with-icon .fi-icon {
  position: absolute;
  width: 18px;
  height: 18px;
  top: 10px;
  right: 10px;
}

.c1.fi-text-input--error .fi-text-input_input {
  border-color: hsl(3,59%,48%);
}

.c1.fi-text-input--error .fi-text-input_statusText {
  color: hsl(3,59%,48%);
}

.c1.fi-text-input--success .fi-text-input_input {
  border-color: hsl(166,90%,36%);
}

.c1.fi-text-input--disabled .fi-text-input_input {
  color: hsl(202,7%,67%);
  background-color: hsl(202,7%,97%);
}

.c1.fi-search-input {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  display: inline-block;
  width: 290px;
}

.c1 .fi-search-input_input-element-container {
  position: relative;
  min-height: 40px;
  border: 1px solid hsl(202,7%,80%);
  border-radius: 2px;
}

.c1 .fi-search-input_input-element-container:focus-within {
  box-shadow: 0 1px 2px 0 rgba(0,53,122,0.1) inset;
}

.c1 .fi-search-input_input-focus-wrapper {
  width: calc(100% - 80px);
}

.c1 .fi-search-input_input-focus-wrapper > input:focus {
  outline-color: hsl(23,82%,53%);
  outline-width: 4px;
  outline-offset: 2px;
}

.c1 .fi-search-input_input-focus-wrapper:focus-within {
  outline: 0;
  position: relative;
}

.c1 .fi-search-input_input-focus-wrapper:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,55%);
  z-index: 9999;
}

.c1 .fi-search-input_input-focus-wrapper:focus-within:not(:focus-visible):after {
  content: none;
}

.c1 .fi-search-input_input-focus-wrapper:focus-within > input:focus {
  outline: none;
}

.c1 .fi-search-input_input {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,80%);
  border-radius: 2px;
  line-height: 1;
  padding-top: 6px;
  padding-bottom: 6px;
  border: 0;
  background-color: hsl(0,0%,100%);
  min-height: 36px;
  margin-top: 2px;
  margin-bottom: 2px;
  min-width: 65px;
}

.c1 .fi-search-input_input::-webkit-input-placeholder {
  font-style: italic;
}

.c1 .fi-search-input_input::-moz-placeholder {
  font-style: italic;
}

.c1 .fi-search-input_input:-ms-input-placeholder {
  font-style: italic;
}

.c1 .fi-search-input_input::placeholder {
  font-style: italic;
}

.c1 .fi-search-input_label-p {
  margin-bottom: 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,16%);
}

.c1 .fi-search-input_statusText_container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 .fi-search-input_statusText_container .fi-search-input_statusText {
  margin-top: 5px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
}

.c1 .fi-search-input_hintText {
  display: block;
  color: hsl(0,0%,16%);
  margin-bottom: 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 .fi-search-input_button-clear {
  position: absolute;
  top: 0px;
  right: 40px;
  height: 20px;
  width: 20px;
  border: 0;
  margin: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  fill: hsl(202,7%,40%);
}

.c1 .fi-search-input_button-clear:focus {
  outline: 0;
  position: relative;
}

.c1 .fi-search-input_button-clear:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,55%);
  z-index: 9999;
}

.c1 .fi-search-input_button-clear:focus:not(:focus-visible):after {
  content: none;
}

.c1 .fi-search-input_button-clear:focus {
  position: absolute;
}

.c1 .fi-search-input_button-clear_icon {
  width: 14px;
  height: 14px;
  fill: hsl(202,7%,67%);
}

.c1 .fi-search-input_button-search {
  position: absolute;
  top: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 40px;
  right: 0px;
  width: 40px;
  border-radius: 0 2px 2px 0;
  background-color: hsl(212,63%,45%);
}

.c1 .fi-search-input_button-search:focus {
  outline: 0;
  position: relative;
}

.c1 .fi-search-input_button-search:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,55%);
  z-index: 9999;
}

.c1 .fi-search-input_button-search:focus:not(:focus-visible):after {
  content: none;
}

.c1 .fi-search-input_button-search:focus {
  position: absolute;
}

.c1 .fi-search-input_button-search-icon {
  width: 18px;
  height: 18px;
  fill: hsl(0,0%,100%);
}

.c1 .fi-search-input_button-search-empty {
  background-color: transparent;
}

.c1 .fi-search-input_button-search-empty .fi-search-input_button-search-icon {
  fill: hsl(202,7%,67%);
}

.c1.fi-search-input--error .fi-search-input_input {
  border-color: hsl(3,59%,48%);
}

.c1.fi-search-input--error .fi-search-input_button-clear {
  border-color: hsl(3,59%,48%);
}

.c1.fi-search-input--error .fi-search-input_button-search {
  border-color: hsl(3,59%,48%);
}

.c1.fi-search-input--error .fi-search-input_statusText {
  color: hsl(3,59%,48%);
}

.c1.fi-search-input--disabled .fi-search-input_input {
  color: hsl(202,7%,67%);
  background-color: hsl(202,7%,97%);
}

.c1.fi-search-input--disabled {
  cursor: not-allowed;
}

<div
  class="c0 c1 fi-search-input"
>
  <label
    class="c2 fi-search-input_label"
  >
    <p
      class="c3 fi-paragraph fi-search-input_label-p"
    >
      Test search input
    </p>
    <div
      class="c0 fi-search-input_statusText_container"
    >
      <div
        class="c0 fi-search-input_input-element-container"
      >
        <div
          class="c0 fi-search-input_input-focus-wrapper"
        >
          <input
            aria-invalid="false"
            class="c4 fi-search-input_input"
            data-testid="textinput"
            id="test-id"
            type="text"
          />
        </div>
        <div
          class="c0 fi-search-input_button-search fi-search-input_button-search-empty"
          role="button"
          tabindex="-1"
        >
          <svg
            aria-label="Search"
            class="fi-icon c5 fi-search-input_button-search-icon"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M9 2C5.14 2 2 5.141 2 9.001 2 12.86 5.14 16 9 16s7-3.14 7-6.999C16 5.141 12.86 2 9 2m7.029 12.615l7.678 7.678a.999.999 0 11-1.414 1.414l-7.678-7.678A8.957 8.957 0 019 18c-4.962 0-9-4.037-9-8.999C0 4.038 4.038 0 9 0s9 4.038 9 9.001a8.955 8.955 0 01-1.971 5.614z"
            />
          </svg>
        </div>
      </div>
    </div>
  </label>
</div>
`;
